Generate constants directly in CreateDebugInlinedAt

greg-lunarg opened the pull request:
Using the ConstantManager attempts to regenerate the DefUseManager
which is not valid during inlining.

jaebaek commented on source/opt/debug_info_manager.cpp:
```diff
@@ -194,9 +194,20 @@ uint32_t DebugInfoManager::CreateDebugInlinedAt(const Instruction* line,
+    // of DefUseManager which is invalidated for inlining. The extra
+    // constants that may be generated here is likely not significant
+    // and will likely be cleaned up in later passes.
     if (line_number_type == spv_operand_type_t::SPV_OPERAND_TYPE_ID) {
```
To avoid creating the same constant multiple times, I would use the constant manager when it is available.
Can we do the following:
```
if (context()->AreAnalysesValid(IRContext::Analysis::kAnalysisConstants & IRContext::Analysis::kAnalysisDefUse)) {
        uint32_t line_id =
          context()->get_constant_mgr()->GetUIntConst(line_number);
} else {
  // create it directly
}
```

greg-lunarg replied:
Well, it certainly is correct, although I don't know if it will ever have any effect. That is because CreateDebugInlinedAt is currently only called from out of inlining which invalidates DefUse almost immediately. I am not sure this function will ever be called from anywhere but the inliner.

If you still want it I will make the change. Let me know.

s-perron reviewed (changes requested):
Tests should be added to https://github.com/KhronosGroup/SPIRV-Tools/blob/e6e77dbdfa799b3be2faa12f4dd8a16f6e3b4cc7/test/opt/debug_info_manager_test.cpp.

jaebaek commented on source/opt/debug_info_manager.cpp:
```diff
@@ -146,6 +146,25 @@ void DebugInfoManager::RegisterDbgDeclare(uint32_t var_id,
 
+// Create new constant directly into global value area, bypassing the
+// Constant manager. This is used when the DefUse or Constant managers
+// are invalid and cannot be regenerated due the module being in an
```
.. cannot be regenerated **due** the module being ..

--> .. cannot be regenerated **due to** the module being ..

?

greg-lunarg replied:
got it.

jaebaek reviewed (commented):
LGTM. I am not sure how we can add a unit test for this.
Maybe calling `DebugInfoManager::CreateDebugInlinedAt()` after disabling DefUse and Constant analysis?

greg-lunarg commented:
@s-perron @jaebaek I have added a test, but I have added it to inline_test.cpp since that is the pass during which the failure occurred and so it was trivial to replicate the test conditions.

s-perron commented:
> I am not sure how we can add a unit test for this.

The idea is that this function should not build the analysis if it was not already available.  The test can call the function that was changed with a much smaller spir-v module.  Create an inlined at instruction, and expect that the analysies are still invalid.  They you can do another test where the constant manager is build before creating the instruction and then expect that it is valid afterwards.

I would much prefer the more focused unit test.

greg-lunarg commented:
I am not sure what is going on with these failures. They appear to be due to github failing to clone repos. I am not sure what to do about that.

greg-lunarg commented:
Looks like a recent change to glslang build has broken shaderc smoke test :(
